# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/angle/checkout/src/libANGLE/renderer/ShaderImpl.h
# Commit: 13282d7a47a5
# Full Hash: 13282d7a47a57a73be8d060efd9645b4d642f03e
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2019-09-08 09:46:52
# Regressor Bug: 1578576
# File Overlap Count: 4
# Description:
#   Bug 1578576 - Part 2: Update to ANGLE 3865 r=jgilbert
#   
#   Depends on D44561
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D44579
# ==============================================================================

diff -r 3064469c073d -r 13282d7a47a5 gfx/angle/checkout/src/libANGLE/renderer/ShaderImpl.h
--- a/gfx/angle/checkout/src/libANGLE/renderer/ShaderImpl.h	Sat Sep 07 13:36:46 2019 +0000
+++ b/gfx/angle/checkout/src/libANGLE/renderer/ShaderImpl.h	Sat Sep 07 13:38:36 2019 +0000
@@ -9,12 +9,42 @@
 #ifndef LIBANGLE_RENDERER_SHADERIMPL_H_
 #define LIBANGLE_RENDERER_SHADERIMPL_H_
 
+#include <functional>
+
 #include "common/angleutils.h"
 #include "libANGLE/Shader.h"
+#include "libANGLE/WorkerThread.h"
+
+namespace gl
+{
+class ShCompilerInstance;
+}  // namespace gl
 
 namespace rx
 {
 
+using UpdateShaderStateFunctor = std::function<void(bool compiled, ShHandle handle)>;
+class WaitableCompileEvent : public angle::WaitableEvent
+{
+  public:
+    WaitableCompileEvent(std::shared_ptr<angle::WaitableEvent> waitableEvent);
+    ~WaitableCompileEvent() override;
+
+    void wait() override;
+
+    bool isReady() override;
+
+    virtual bool getResult() = 0;
+
+    virtual bool postTranslate(std::string *infoLog) = 0;
+
+    const std::string &getInfoLog();
+
+  protected:
+    std::shared_ptr<angle::WaitableEvent> mWaitableEvent;
+    std::string mInfoLog;
+};
+
 class ShaderImpl : angle::NonCopyable
 {
   public:
@@ -23,22 +53,20 @@
 
     virtual void destroy() {}
 
-    // Returns additional sh::Compile options.
-    virtual ShCompileOptions prepareSourceAndReturnOptions(const gl::Context *context,
-                                                           std::stringstream *sourceStream,
-                                                           std::string *sourcePath) = 0;
-
-    // Uses the GL driver to compile the shader source in a worker thread.
-    virtual void compileAsync(const std::string &source, std::string &infoLog) {}
-
-    // Returns success for compiling on the driver. Returns success.
-    virtual bool postTranslateCompile(gl::ShCompilerInstance *compiler, std::string *infoLog) = 0;
+    virtual std::shared_ptr<WaitableCompileEvent> compile(const gl::Context *context,
+                                                          gl::ShCompilerInstance *compilerInstance,
+                                                          ShCompileOptions options) = 0;
 
     virtual std::string getDebugInfo() const = 0;
 
     const gl::ShaderState &getData() const { return mData; }
 
   protected:
+    std::shared_ptr<WaitableCompileEvent> compileImpl(const gl::Context *context,
+                                                      gl::ShCompilerInstance *compilerInstance,
+                                                      const std::string &source,
+                                                      ShCompileOptions compileOptions);
+
     const gl::ShaderState &mData;
 };
 